### ISSUE TITLE
Running test on bitbar cloud

### ISSUE THREAD

**Issue by robert-altom** [robert-altom/test#798]
Hey, i tried to follow steps of @ru.cindrea to run test on bitbar device farm. Sadly, i failed. Is there any chance for detailed guide/introduction how to prepare environment and what which scripts should contain?

---
<sub>You can find the original issue from GitLab [here](https://gitlab.com/altom/altunity/altunitytester/-/issues/92).</sub>

**Comment by robert-altom** [comment]
We added a section for this to our documentation: https://altom.gitlab.io/altunity/altunitytester/pages/tester-with-cloud.html

**Comment by robert-altom** [comment]
Thanks, @ru.cindrea. I'm trying to explore AWS device farm with AltUnityTester using an Appium Java TestNG package.
I keep getting java.net.ConnectException: Connection refused error on the following line: 
altUnityDriver = new AltUnityDriver("127.0.0.1", 13000);

Should I mention some other IP?

The appium server on AWS runs on 0.0.0.0:4723.

**Comment by robert-altom** [comment]
@tushar3 Yes, I believe SauceLabs need to open another port for AltUnityTester before the tests will work, and you cannot do it yourself like you can with BitBar. 

BitBar have the "Appium Server Side execution" projects where you get a docker container with a device connected to it, and you can do whatever you want inside it - including forwarding the port for AltUnityTester, which is great. 

AWS device farm has a similar implementation to BitBar, I managed to get AltUnityTests to run there but I need to update the docs for that, they are no longer up to date...

**Comment by robert-altom** [comment]
Great that it works @PvanderLaan 

If you plan to try it with iOS as well, the current version has a bug (#107) that is still open and it kills the connection to the Webdriver Agent. To work around that, I've added the following to my test, right after creating the Appium and the AltUnityDriver:

```
        try:
            subprocess.Popen(['iproxy', '8100', '8100'])
        except:
            print('tried to forward WebdriverAgent')

```

**Comment by robert-altom** [comment]
Thanks a lot, Ru! I changed the version and now the test works on the BitBar cloud :) Hopefully I can soon instruct the devs here to implement AltUnityTester and we can do detailed tests on the cloud with our games.

**Comment by robert-altom** [comment]
@PvanderLaan try with the latest version of altunityrunner in your requirements.txt:
`altunityrunner==1.2.2`

**Comment by robert-altom** [comment]
@ru.cindrea I'm also trying to do a cloud test on BitBar. I've come to the point where I can run Appium tests on the Bitbar cloud, but I'm stuck at the step of opening the tcp connection for UnityAltTester to work.

I added 'adb forward tcp:13001 tcp:13000' to the run-test.sh (see attached files) but I get this response in console.log:

````
======================================================================
ERROR [0.000s]: setUpClass (__main__.SampleAppiumTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "BitbarSampleAppTest.py", line 24, in setUpClass
    self.altdriver = AltrunUnityDriver(self.driver, 'android')
  File "/usr/local/lib/python2.7/dist-packages/altunityrunner/runner.py", line 96, in __init__
    raise Exception('AltUnityServer not running on port ' + str(TCP_PORT) + ', did you run ``adb forward tcp:' + str(TCP_PORT) + ' tcp:13000`` or ``iproxy ' + str(TCP_PORT) + ' 13000``?')
Exception: AltUnityServer not running on port 13001, did you run ``adb forward tcp:13001 tcp:13000`` or ``iproxy 13001 13000``?

----------------------------------------------------------------------
Ran 0 tests in 94.378s
````

Do you have any tips for me or should I contact BitBar?

Scripts used:
[bitbar_adventure_test05.zip](https://gitlab.com/Altom/AltUnity/AltUnityTester/uploads/cc9887a4376c3579e352eba275fb8abf/bitbar_adventure_test05.zip)

**Comment by robert-altom** [comment]
@ru.cindrea  @robert\_poienar  I'm trying a similar setup with SauceLabs - TestObject. Have you guys explored it?

I'm going to try using their [Pre-Run Executables](https://wiki.saucelabs.com/display/DOCS/Setting+Up+Pre-Run+Executables) to pass the port forward command before the test.

Update: I tried the above mentioned method and also tried to use the driver.executeScript command from appium to open the port forward and got an error that the sauce labs appium server does not have the relaxed security flag set in order to run shell commands. I've raised an issue with them to see if this is possible.

**Comment by robert-altom** [comment]
@ru.cindrea Thanks, i have followed the same steps in above doc, but keep getting "iproxy: no process found" when we initialise "AltrunUnityDriver".

**Comment by robert-altom** [comment]
To run my tests on Bitbar cloud, all I did was:
1. follow their example from here: https://github.com/bitbar/bitbar-samples/tree/master/appium/server-side/python 
2. add the python binding installation in the requirements.txt script:altunityrunner==1.2.2
3. add the adb forward command in “run-scripts.sh”, just before starting Appium:
`adb forward tcp:13001 tcp:13000`

I haven't yet tried iOS, I assume one might need to install iproxy? I will check at some point and try it out. 

@Mdirfan25 I've tried once to run Android (not iOS) in AWS farm and I was able to do it. I started writing the instructions for it and it's still an open MR (!20) but I haven't had time to fix the merge conflicts after the latest update to AltUnityTester... however, the Readme instructions I created then can be found here: https://gitlab.com/altom/altunitytester/blob/cce7802d1cc80b4975ad58ba3bafdd987ad1387a/aws-device-farm-example/README.md 

Maybe you'll find these useful?
